### DISCUSSION TITLE
devDependencies with Native libraries

### DISCUSSION THREAD

**Discussion by rigor789** [NativeScript/rfcs#3] [1. Ideas & Discussions]
Currently, native code from `dependencies` is picked up by NativeScript and built with the app, however `devDependecies` are not handled the same way.

Sometimes a `devDependency` could require a native module, but only when running the app in debug/dev mode.

NativeScript should include native code from `devDependencies` unless `--release` is passed.

The developer is responsible for guarding the code that invokes native code from `devDependencies`.

An example where this would be useful:
`nativescript-vue-devtools` uses `nativescript-toasty` and `nativescript-socketio`, both have native code. Devtools is a `devDependency` however, to use them - the developer has to install `nativescript-toasty` and `nativescript-socketio` as regular dependencies. This works fine - however for a `--relase` build, the developer has to manually remove these dependencies from `package.json`.

**Comment by farfromrefug** [comment]
I see one issue with this: using a fork of N

myself is using a fork of N until some of my PRs are merged.

To a good working project here is what i want (and have):
* use imports from `@nativescript/core`
* use `tsconfig.json` paths options to have "click to definition" to go to my fork module in node_modules (that s just a plus for me)
* use webpack to alias `@nativescript/core` to my fork module

Now to make all that work i need to have both `@nativescript/core` and my forks as dependencies. However they cant be both picked up because they have native deps => build fail because of duplicate libs/symbols
So my trick to get this to work was simple. Put `@nativescript/core` in `devDependencies`. Exactly because it would not get picked up by N!

It might seem like a "corner case" but i dont think it is. It is very important for me (and maybe other once they see that trick!) to be able code "with nativescript/core" but be able to use a fork for "some time". If you look at it from the client side it means he can be confident his code is ready to "switch" back to the N core while being able to profit from PRs before they are merged.

So what you propose would break actually all my apps :s . Their might be to go around this but i dont see it right now. Need to think about it

**Comment by rigor789** [comment]
@farfromrefug I think we could utilize `optionalDependencies` for your use-case: https://docs.npmjs.com/cli/v6/configuring-npm/package-json#optionaldependencies

**Comment by NathanaelA** [comment]
I'm thinking this would complicate things, especially for Playground and would cause issues with any test resources found in devDeps.   devDeps are used for anything "build" time related with NativeScript, i.e. all the build (or development) tooling.   You aren't supposed to have any "build-able" resources in the development Deps for NS.    

But I understand your problem, not sure a good solution as we do need to keep the development tooling resources (currently the devDeps) separate as is a LOT of npm modules and a lot of them have a lot of extra crap that we don't want accidentally merged into the app.

**Comment by farfromrefug** [comment]
What about a simple option to filter packages in production? Like those would not looked in for native deps. 
We could even add a hook for it so that plugin can add to it (if they now what they are doing ;))

**Comment by rigor789** [comment]
> I'm thinking this would complicate things, especially for Playground 

I don't see how this would affect the playground

>  devDeps are used for anything "build" time related with NativeScript, i.e. all the build (or development) tooling

Exactly the point, in this case vue-devtools is a tool for developing! 

>  You aren't supposed to have any "build-able" resources in the development Deps for NS.

Indeed - that's what I feel like should change from "not supposed" to "you can... (but only for development)"

> a LOT of npm modules and a lot of them have a lot of extra crap that we don't want accidentally merged into the app.

That is true, however the only "crap" I can think of is if the deps had an `android/ios` folder that has the required structure to be picked up by NS - other code like `js` or `json` (or anything else) would never be magically in your bundle/app unless you explicitly import/require it.

But I do see this being a risky change, as not everyone understands the difference between `dependencies`/`devDependencies` and could end up installing ns plugins to `devDependencies`, seeing them work fine locally, but after a `--relase` build the app would crash. 

> We could even add a hook for it so that plugin can add to it.

The more I think of this, the more I think would be the better solution. A dependency (both dev and regular) could have a hook that "enqueues" itself (or it's own dependencies) as needing the native dependencies built into the app. In my particular example, `nativescript-vue-devtools` would add `nativescript-toasty` and `nativescript-socketio` as native dependencies if in a dev build. The CLI would need to de-dupe these, since either of the plugins could already be in `dependencies` we don't want to build/include them twice (and fail).

---

Thanks for the input @NathanaelA and @farfromrefug - I think the hook approach would be safer.
